src/redisearch_rs/varint_bencher/src/c_varint.rs

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Implementation of C varint wrapper types that provide safe Rust interfaces
+//! to the C varint functions for benchmarking comparisons.
+
+use crate::FieldMask;
+use crate::ffi;
+
+/// A safe wrapper around the C VarintVectorWriter implementation.
+#[repr(transparent)]
+pub struct CVarintVectorWriter(*mut ffi::VarintVectorWriter);
+
+impl CVarintVectorWriter {
+    /// Create a new C-backed varint vector writer with the given capacity.
+    #[inline(always)]
+    pub fn new(cap: usize) -> Self {
+        // SAFETY: NewVarintVectorWriter is a valid C function that creates a new vector writer
+        let ptr = unsafe { ffi::NewVarintVectorWriter(cap) };
+        assert!(!ptr.is_null(), "Failed to create C VarintVectorWriter");
+        Self(ptr)
+    }
+
+    /// Write a varint to the vector writer.
+    /// Returns the number of bytes written.
+    #[inline(always)]
+    pub fn write(&mut self, value: u32) -> usize {
+        // SAFETY: VVW_Write is a valid C function and self.0 is a valid VarintVectorWriter pointer
+        unsafe { ffi::VVW_Write(self.0, value) }
+    }
+
+    /// Get the number of values written to the vector.
+    #[inline(always)]
+    pub fn count(&self) -> usize {
+        // SAFETY: self.0 is a valid VarintVectorWriter pointer, accessing nmemb field
+        unsafe { (*self.0).nmemb }
+    }
+
+    /// Get the total number of bytes written.
+    #[inline(always)]
+    pub fn bytes_len(&self) -> usize {
+        // SAFETY: self.0 is a valid VarintVectorWriter pointer, accessing buf.offset field
+        unsafe { (*self.0).buf.offset }
+    }
+
+    /// Get a pointer to the internal byte data.
+    /// The data is valid until the next write operation or until the writer is dropped.
+    #[inline(always)]
+    pub fn bytes_data(&self) -> *const u8 {
+        // SAFETY: self.0 is a valid VarintVectorWriter pointer, accessing buf.data field
+        unsafe { (*self.0).buf.data as *const u8 }
+    }
+
+    /// Truncate the vector writer.
+    /// Returns the new capacity.
+    #[inline(always)]
+    pub fn shrink_to_fit(&mut self) -> usize {
```

Comment:
Are all the methods used / needed?

Reply:
Good question. I'll check and remove ones that aren't used.